### ISSUE TITLE
No way to tell a "suggestion" from a "tool"

### ISSUE THREAD

**Issue by justcla** [dotnet/roslyn#28241]
The lightbulb menu can show a mix of suggestions and tools. Suggestions are things that VS is proposing you seriously consider as a means of fixing a warning or error. Tools are handy things that can change your code - but do not address any known issues directly.

The menu can contain a mix of these items, but there is no easy way for a user to tell which are suggestions and which are tools.

Here's an example:
One of these is a suggestion, one of them is not. Do you know which is which?

![image](https://user-images.githubusercontent.com/17131343/42184837-b87ab392-7dfb-11e8-9da9-5f14caad9516.png)

**Comment by CyrusNajmabadi** [comment]
Note: In this case, it looks like both of these should be tools.  I don't see anything in the code that would indicate that either is a suggestion.  So it seems buggy (likely on roslyn's side) that we have a lightbulb here.

@sharwell My guess is that we're not properly mapping an analyzer set to none/silent to a refactoring/tool treatment in the UI.  Is this something you were planning to do as part of your work here?

**Comment by CyrusNajmabadi** [comment]
Note: if you actually had suggestion-dots, or a warning/error squiggle for "add accessibility modifiers" then i would expect this to be a lightbulb.  So, for the purposes of this discussion can you add *another* image that shows that?  You'll have to change your severity for the option to require accessibility modifiers.

Thanks!

**Comment by justcla** [comment]
> it looks like both of these should be tools.

Wouldn't it be nice if I could tell?
That's my whole point.

**Comment by justcla** [comment]
When I add null checks and add accessibility modifiers, the icon becomes a screwdriver.

![image](https://user-images.githubusercontent.com/17131343/42186270-e1a58ad0-7e00-11e8-9b29-90417edb4d13.png)

From that I deduced that Change Signature was a tool and the others were suggestions.

**Comment by justcla** [comment]
If there were icons beside each item, it would be very clear which ones are real suggestions.

![image](https://user-images.githubusercontent.com/17131343/42186402-69ab1ca6-7e01-11e8-9c18-9dbd7686bda6.png)

**Comment by justcla** [comment]
I do think it's weird that lightbulb actions are surfacing where there are no squiggles. Why is there no squiggle for "Add accessibility modifiers"? It's not showing up in the Error List, either. Perhaps it should be a screwdriver too.
Same too for Add null check, which surfaced without any squiggly.

**Comment by CyrusNajmabadi** [comment]
> From that I deduced that Change Signature was a tool and the others were suggestions.

That's not a logical deduction.  We may or may not offer toosl depending on what your code looks like, just as we may or may not offer suggestions based on what your code looks like.  Once you change your code, the set of offered things (be they suggestions or tools) may necessarily change.

**Comment by justcla** [comment]
I'll try to provide some other examples where the menu is full of items - some suggestions, some not - and I can't tell which is which.

**Comment by CyrusNajmabadi** [comment]
> Well whatever it is, it confused me.

All i'm saying is that i want to address the confusion with the *right* fix.  In my very first post i called this out and brought up the two primary issues i'm seeing.  I'm trying to address this in a systematic way as i believe there are two issues going on.

1. Some things that shuold be surfaced as tools are surfaced as suggestions.  This is your first image.
2. Even when things should be surfaced as suggestion, it will be unclear which offerings are actually hte suggestions and which are tools.  That's why i asked: "So, for the purposes of this discussion can you add another image that shows that? You'll have to change your severity for the option to require accessibility modifiers."

I'm basically on your side here.  But i'm just trying to keep the issue as confusion free as possible :)

**Comment by justcla** [comment]
Aside from the details of specific fixes (like those mentioned above), what do you think of my made-up image that shows a mix of bulbs and screwdrivers?

**Comment by CyrusNajmabadi** [comment]
> I'll  try to provide some other examples where the menu is full of items -  some suggestions, some not - and I can't tell which is which.

Yes.  That's what i was asking for when i said:

> So, for the purposes of this discussion can you add another image that shows that? You'll have to change your severity for the option to require accessibility modifiers

I think this will demonstrate the problem very well.

**Comment by CyrusNajmabadi** [comment]
> Aside from the details of specific fixes (like those mentioned above), what do you think of my made-up image that shows a mix of bulbs and screwdrivers?

With the image you provided, it does not make sense.  Since all of those should be tools.  **However**, i think it would make sense if you also had an image that had suggestion-dots, or squiggles in the appropriate location.

**Comment by Therzok** [comment]
Note: in VSMac, we put code fixes at the top of the list, a separator then list all the refactorings.

I know Roslyn has a MRU cache, not sure how that would behave in that case. We consider code fixes as more important than refactorings by default.

**Comment by justcla** [comment]
If the items were in strict order, a separator would be a good solution.

But we like to mix things up (ie. Rename refactoring), so I don't know if we can rely on a fixed ordering.

**Comment by Therzok** [comment]
To expand, we do the following:
```
MRU code fixes
Separator
MRU refactorings
```

So it's not a fixed order per se, it's just giving refactorings lesser priority due to the fact that they're not really the reason a user went to a squiggle (the equivalent of VS lightbulb) menu in code

**Comment by CyrusNajmabadi** [comment]
Note: we have a PR for roslyn as well do make it so that if the user has a selection, then a refactoring that operates on that suggestion is given higher priority.  After all, if i select some subsection of squiggled code so that i can extract it, it's hugely annoying to have arbitrary suggestions take precedence of the code i'm specifically demarcating with my selection to operate on.

--

yaay more complexity :)

**Comment by justcla** [comment]
Here we go.

![image](https://user-images.githubusercontent.com/17131343/42187497-23e0c15e-7e05-11e8-9ecf-12aeeb8fbc69.png)

There's some squiggles, and a bunch of items in the menu.

Can you tell me which ones will address problems and which ones will definitely not?

I know from experience that the last item will NOT fix any of the issues being raised. It is not a suggestion.

**Comment by justcla** [comment]
If it looked like this, it would be clear.

![image](https://user-images.githubusercontent.com/17131343/42187772-16fb6182-7e06-11e8-8351-47b88c096f48.png)

**Comment by CyrusNajmabadi** [comment]
@justcla Yup.  I think we could def try this out.  I'd be interested in the experience.

**Comment by CyrusNajmabadi** [comment]
> Note: we have a PR for roslyn as well do make it so that if the user has a selection, then a refactoring that operates on that suggestion is given higher priority. After all, if i select some subsection of squiggled code so that i can extract it, it's hugely annoying to have arbitrary suggestions take precedence of the code i'm specifically demarcating with my selection to operate on.

oops.  I misspoke.  We don't have a PR for this.  We already accepted this change back in april :)

**Comment by CyrusNajmabadi** [comment]
> I know Roslyn has a MRU cache, not sure how that would behave in that case. We consider code fixes as more important than refactorings by default.

Roslyn does not use any sort of MRU cache for lightbulbs.  Our logic is pretty complex, but it can be broken down to a few cases:

1. No selection.  In that case, fixes first, then tools.
2. Selection.  in that case, tools first, then fixes.  

Note: we def have some issues where some items are misclassified.  i.e. some things are classified as fixes when they should be tools (and possibly vice versa).  Roslyn def needs to fix that.

Now, Given the above, we also have things like:

1. Allow providers to give a high/medium/low pri to the tools/suggestion.  In hte tool and suggestion groups, things will be ordered accordingly based on those buckets.
2. Allowing providers to configure relative ordering per their section.  i.e. tools from a specific provider can order themselves against other tools against another provider.  I believe high/medium/low takes effect first, then this takes effect next.  This allows one fix to say "we're less important than 'add using', it should always come first".  This is the vs MEF 'Order' concept.

Finally, position of caret on line comes into play and distance of the fixes from that caret is assesses.  However, without looking at the code, i'm not sure where that is done in this list i've given so far.  

Sooooooo much complexity....

**Comment by CyrusNajmabadi** [comment]
Ok.  We do the sorting-by-distance last.  Which i *think* then means it actually has the highest priority.  Because it will effectively take all the sorting/grouping we've done so far, and then actually reprioritize everything based on distance.  So, take the above algorithm, but add a section at the top which is: first group things by distance.  Things that are closer are higher, things that are further away are lower.  Then, for any distance group, apply the above other rules.

Note: i think this is *very* appropriate.  When i move my caret onto something to manipulate, i don't want something 'higher severity', but 100 chars away on the line interfering with the thing my caret is on now and which i want to manipulate.

--

Finally, on top of *all* of that, Roslyn also has a fairly interesting heuristic around 'inlining' lightbulb items.  For example, you'll note in Justin's picture that '3' items are ones taht have further items to the right you can select:

![image](https://user-images.githubusercontent.com/4564579/42188840-7314df12-7e0a-11e8-822f-4bed40f37110.png)

(i.e. ```Generate variable 'argv'```)

Now, this is how we display things if there are a lot of items in the list.  That's because we want to give users a clear set of distinct top level choices they can make, and we don't want the list to be incredibly overcrowded. However, if there are actually very few top level items in teh list (i.e. the lightbulb would only have ```Generate variable 'argv'```) then we'll actually 'inline' all the sub-item suggestion directly in the list so that you get:

![image](https://user-images.githubusercontent.com/4564579/42188894-af90b2ea-7e0a-11e8-8b61-b8560a0afc3f.png)

This approach allows us to both keep the list from getting majorly overloaded (if everythign were inlined), without also making things super annoying (i.e. "why do i always have to select a subitem when there are only 4 things to chose from??").

**Comment by CyrusNajmabadi** [comment]
Yup yup.  I'll see what i can do with @sharwell to improve things in this area!

**Comment by CyrusNajmabadi** [comment]
I think we'll likely want an 'experiment' around the iconography.  Tagging @kuhlenh to make her aware of these discussions.

**Comment by justcla** [comment]
On the issue of sorting by closest distance, I find this odd.
I totally get that when your caret is ON a squiggle, or your selection would trigger an action (like extract variable), those actions should be ranked higher than other actions on that line that are not EXACTLY where the caret/selection is. That is important, as you mentioned, so that you can navigate to the point and trigger the first item in the menu.
But I think the flaw is in using "nearest match". When your caret is between two squiggles, you will get a different order depending on how far to the left or right the caret is between the items.
My view is, unless you are actually ON the squiggle, it should not be artificially surfaced to the top.

This would provide better consistency and predictability. I should be able to get used to knowing the order items will appear in; whether I press Home or End, the items in the list should be in the same order - unless I happen to have navigated onto a squiggle or made a selection that triggers specific actions.

Slightly off-topic. Should I raise this as a separate issue?

**Comment by CyrusNajmabadi** [comment]
> But I think the flaw is in using "nearest match". When your caret is between two squiggles, you will get a different order depending on how far to the left or right the caret is between the items.
My view is, unless you are actually ON the squiggle, it should not be artificially surfaced to the top.

This is challenging because the user may often not be (exactly) on an item, but definitely considers themselves at that item.  Consider something like  ```foo();``` (where the user is after teh semicolon).  The squiggle will be on ```foo()``` (because that's where the error is is), but the user won't be on that squiggle.

The language and analyzers are full of these cases.  

> This would provide better consistency and predictability. I should be able to get used to knowing the order items will appear in; whether I press Home or End, the items in the list should be in the same order

I disagree completely.  If i'm at the front of a line, and there is an error one character in front of me, then that is far more important than if i'm 100 characters in and there's an error one character in front of that.

> Slightly off-topic. Should I raise this as a separate issue?

You're welcome to (always!) :).  However, i would strongly ask that we discuss these things and get background on it.  As i've mentioned previously, these have all be extremely intentional decisions based on over 10+ years of experience with actual analyzers and issues being surfaced in the code.

**Comment by justcla** [comment]
Interesting thoughts.

I guess I feel differently. I expect accuracy with tools - putting the effort on me to type correctly and place the caret correctly. (Hence why I don't want spell check about Generate variable).
Various tools help with this kind of "accurate typing/placement" behavior. For instance, a good "Go to next issue" command with shortcut can easily move the caret onto the next squiggle.

But I hear that there was some previous experimentation done in this area. I'm open to reviewing it and seeing whether the issues of the day are still relevant today, especially with the many changes that have occurred to the product over time.

One of the worst cases of this NOT happening, when it should is when my caret is on the right of a squiggle. Unless it's the last char on the line, it won't open the relevant suggestions. So while the caret placement thing causes problems with inconsistency, it also fails to provide contextual information when it should. (Probably the topic for a separate issue - I expect there's one already for this.)

**Comment by CyrusNajmabadi** [comment]
> But I hear that there was some previous experimentation done in this area. I'm open to reviewing it and seeing whether the issues of the day are still relevant today, especially with the many changes that have occurred to the product over time.

I don't see anything that would have changed any of these issues.  IMO the onus would be on anyone wanting to make changes here to instead have to provide strong demonstrative results that show that these changes are actually an improvement.  These decisions in the past were not made haphazardly.  They were made by a continued focus on sweating these small details and collecting and continually iterating on lots of feedback over the years.

Most particularly, when we get this stuff wrong, people tend to notice and get quite upset.  Not taking location into account when we changed to the left-margin-lightbulb was a good example of that.  For years people had been used to being able to just do  ```... code ... Foo();```, hit ctrl-dot and generate their new 'Foo' member.  When left-margin-lightbulbs came online and we only prioritized an issue if you were on it, this *was* a serious issue.  People felt this immediately and didn't like that we were showing them fixes for things far away in code when they actually wanted fixed for what they had just been editing.

**Comment by justcla** [comment]
Lots to discuss on sorting by caret location. Holding over to discuss this on another issue.

**Comment by sharwell** [comment]
@mikadumont Perhaps we should group this with the new UX that we already need to create to support filtering the list by typing?

**Comment by sharwell** [comment]
Design review conclusion: we haven't seen specific feedback regarding the ability to see all of the "fixes" from all of the "refactorings" when expanding the light bulb. However, we've heard feedback numerous times regarding the difficulty of quickly identifying "proper fixes for specific diagnostics I see". We'll close this item as Won't Fix, but leave the other items open to find new ways to draw users to recommended/suggested changes from code fixes.
